zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ZookeeperConnectorBuilder.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.zookeeper;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.NodeCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.TreeCache;
+import org.apache.curator.framework.recipes.leader.LeaderLatch;
+import org.apache.curator.framework.recipes.leader.LeaderSelector;
+import org.apache.curator.framework.recipes.nodes.GroupMember;
+import org.apache.curator.framework.recipes.nodes.PersistentNode;
+import org.apache.curator.framework.recipes.nodes.PersistentTtlNode;
+import org.apache.curator.framework.recipes.queue.DistributedDelayQueue;
+import org.apache.curator.framework.recipes.queue.DistributedIdQueue;
+import org.apache.curator.framework.recipes.queue.DistributedPriorityQueue;
+import org.apache.curator.framework.recipes.queue.DistributedQueue;
+import org.apache.curator.framework.recipes.shared.SharedCount;
+
+import com.google.common.collect.Lists;
+
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerListener;
+import com.linecorp.armeria.server.ServerListenerBuilder;
+
+/**
+ * Provides armeria-zookeeper connectivity for {@link CuratorFramework} and its high-level abstractions, like
+ * following:
+ *
+ * <p>- Starts given {@link CuratorFramework} instances in specified order when {@link Server} started.
+ * - Stops given {@link CuratorFramework} instances in reverse order when {@link Server} stopping.
+ *
+ * <p>Note: you don't have to start or close the {@link CuratorFramework} instance used in
+ * {@link ZooKeeperUpdatingListenerBuilder}.
+ * */
+public class ZookeeperConnectorBuilder {
```

Comment:
It seems like this class is not really used anywhere else. Could you explain why we need this?

Reply:
This class provides the functionality of #882.

Reply:
Why is this better than starting and stopping `CuratorFramework` in `ZooKeeperUpdatingListener`? `ZooKeeperUpdatingListener` could start/stop `CuratorFramework` if a user specified a connection string, or leave `CuratorFramework` as is if it's provided by a user.

Reply:
Perhaps you meant to refer to this comment: https://github.com/line/armeria/issues/882#issuecomment-364348308

However, I don't see why this rather generic utility is necessary for service discovery. Could you show me a specific use care for this class? If there's no other use case than starting/stopping `CuratorFramework` during service registration, this class could be removed.